NIFI-1678

markap14 opened the pull request:
Refactored NCM & Nodes so that nodes send heartbeats to ZooKeeper and NCM reads heartbeats from ZooKeeper.

mcgilman commented:
Ran into an issue with the PR as is. Specifically, I deployed the contributions to an existing cluster. The existing cluster had already stored state in ZK using the configuration from the state-management.xml. During start up the nodes were unable to heartbeat into ZK because the configuration in state-management.xml had used a username/password. In order for the heartbeats to be stored the ZK connection from nifi.properties would also need to be configured with the same credentials.

As I mentioned in the PR for #301 we should address having duplicate connect string in nifi.properties and state-management.xml for ZK. A short term fix for addressing this issue would be to additionally duplicate the remaining ZK configuration from state-management to nifi.properties. We could take this temporary approach while we discuss a more appropriate long term strategy for ZK configuration prior the first 1.x release. Or alternatively, we could address that issue prior to re-submitting this PR.

markap14 commented:
@mcgilman I agree - we certainly need to address the duplication of configuration. However, I do believe that is a job for another ticket. I went ahead and added username/password configs for zookeeper into the nifi.properties file for now. We can figure out how best to refactor the configuration in a later ticket before 1.0 is released.

I did push a new commit, which I believe addresses the issue that you laid out, though, about auth failures. I also found an edge case that wasn't being handled very well, so I addressed that as well in this latest commit and added a unit test to verify.

markap14 commented:
I updated the PR with a new commit to address an issue that I ran into. When we create 'container' nodes in ZooKeeper whose paths are not sensitive, we will mark the node as wide-open, rather than applying ACLs. This allows us to not have to worry about if we have state management and heartbeats using different ACLs.

Also removed use of username/passwords from ZooKeeper configuration. This was done because ZooKeeper indicates that these should be used only for local testing and not for production. In production, security should be administered via Kerberos. Username and password were always sent over the wire unencrypted. Additionally, we are now using Apache Curator to handle Leader Election, and Curator does not appear to support using digest auth via username/password.
